File: problem_11.rb
class ProductGrid

  private

  attr_accessor :grid
  attr_writer :products

  def initialize(grid)
    self.grid = File.read(grid).split("\n").map do |line|
      line.split(' ').map { |number| number.to_i }
    end
    self.products = lines + columns + diagonals1 + diagonals2
  end

  def lines
    grid.each_with_object([]) do |line, products|
      (0..line.length - 4).each do |i|
        products.push(line[i] * line[i+1] * line[i+2] * line[i+3])
      end
    end
  end

  def columns
    grid.transpose.each_with_object([]) do |line, products|
      (0..line.length - 4).each do |i|
        products.push(line[i] * line[i+1] * line[i+2] * line[i+3])
      end
    end
  end

  def diagonals1
    (0..grid.length - 4).each_with_object([]) do |i, products|
      (0..grid.length - 4).each do |j|
        products.push(grid[i][j] * grid[i+1][j+1] * grid[i+2][j+2] * grid[i+3][j+3])
      end
    end
  end

  def diagonals2
    (0..grid.length - 4).each_with_object([]) do |i, products|
      (0..grid.length - 4).each do |j|
        products.push(grid[i][j] * grid[i+1][j-1] * grid[i+2][j-2] * grid[i+3][j-3])
      end
    end
  end

  public

  attr_reader :products

  def maximum_product
    products.max
  end

end

if __FILE__ == $PROGRAM_NAME
  puts ProductGrid.new('grid_pb_11.txt').maximum_product
end

File: problem_11.py
import numpy as np

def main():
    numbers = grid_to_arrays('grid_pb_11.txt')
    products = lines(numbers) + columns(numbers) + diagonals(numbers)
    print(max(products))

def lines(arrays):
    products = []
    for array in arrays:
        for i in range(len(array)-3):
            products.append(array[i] * array[i+1] * array[i+2] * array[i+3])
    return products

def columns(arrays):
    transposed_arrays = np.array(arrays).T
    return lines(transposed_arrays)

def diagonals(arrays):
    products = []
    for i in range(len(arrays)-3):
        for j in range(len(arrays)-3):
            products.append(arrays[i][j] * arrays[i+1][j+1] * arrays[j+2][i+2] * arrays[i+3][j+3])
    for i in range(len(arrays)-3):
        for j in range(len(arrays)-3):
            products.append(arrays[i][j] * arrays[i+1][j-1] * arrays[i+2][j-2] * arrays[i+3][j-3])
    return products


def grid_to_arrays(txt):
    with open(txt) as file:
        lines = [[int(num) for num in line] for line in [line.rstrip().split(' ') for line in file]]
        return lines

if __name__ == "__main__":
    main()
